Tech Debt (code reuse, tests, documentation)

ofisk opened the pull request.

ofisk committed:
Refactor: Consolidate assessment types and campaign state utilities

Phase 1 & 2 of tech debt cleanup - eliminate code duplication and add comprehensive tests

Code Consolidation:
- Create src/types/assessment.ts with centralized type definitions (UserState, CampaignReadinessSummary, ActionSuggestion, ToolRecommendation)
- Create src/lib/campaign-state-utils.ts with shared getCampaignState() and generateReadinessSummary() functions
- Remove duplicate type definitions from assessment-service.ts and state-analysis-tools.ts
- Update assessment-tools.ts to use shared utilities and fix return types (campaignState instead of overallScore)
- Update suggestion-tools.ts to use shared getCampaignState() utility
- Add educational note about campaign state fluctuations during growth

Campaign State Utilities:
- getCampaignState(): Maps scores to 9 RPG-themed descriptive states
- generateReadinessSummary(): Single source of truth for readiness messaging with growth education
- getCampaignStateDescription(): Provides encouraging context for each state
- getNextMilestone(): Returns specific actionable steps to reach next state with detailed guidance

Test Coverage:
- Add comprehensive AssessmentService tests covering all state boundaries (Fresh Start through Legendary)
- Add CampaignAutoRAG service tests for filter enforcement, shard operations, and R2 path construction
- Test shard approval/rejection flows, staging operations, and legacy format handling
- Validate campaign state mapping at key thresholds (0, 20, 30, 40, 50, 60, 70, 80, 90, 100)
- All 30 tests passing

Benefits:
- Zero duplicate interface definitions across codebase
- Single implementation of getCampaignState logic
- Consistent campaign state messaging for users
- Educational context about score fluctuations during campaign growth
- Comprehensive test coverage for critical assessment flows

ofisk committed:
Add comprehensive integration tests for shard approval and context capture flows

- Create shard approval flow integration test (3 tests passing)
- Create context capture flow integration test (6 tests passing)
- Test complete workflows: upload → staging → approval → search
- Test conversational context capture flow with AI detection
- Test pre-approved context (campaign info, characters, sheets)
- Test shard expansion during approval
- Test approved/rejected shard separation
- Test context updates replacing approved shards
- All 9 integration tests validating critical user workflows

ofisk committed:
Documentation updates and dead code removal

Phase 3 & 4 of tech debt cleanup - improve documentation accuracy and remove unused code

Documentation:
- Add Campaign State System section to SHARD_APPROVAL_SYSTEM.md with state table and fluctuation explanation
- Create comprehensive ASSESSMENT_SYSTEM.md documenting the 9-state system, scoring algorithm, and UX philosophy
- Delete duplicate CAMPAIGN_SNIPPET_FLOW.md (identical to CAMPAIGN_SHARD_FLOW.md)
- Update inline comments in assessment-tools.ts to clarify descriptive states
- Update inline comments in assessment-service.ts with detailed scoring algorithm
- Update inline comments in campaign-autorag-service.ts explaining three-folder system
- Update state-analysis-tools.ts comments to emphasize user-friendly output

Dead Code Removal:
- Remove unused CampaignReadinessAssessment interface from campaign-service.ts
- Remove unused assessCampaignReadiness() method (75 lines) - superseded by AssessmentService
- Validate remaining TODOs are still accurate (semantic search, AI-assisted features)

Benefits:
- Documentation now accurately reflects campaign state system
- Clear explanation of why states fluctuate (encourages healthy growth)
- Eliminated 90+ lines of dead code
- Better LLM context with updated comments explaining system design

aniham commented on docs/ASSESSMENT_SYSTEM.md:
```diff
@@ -0,0 +1,296 @@
+## Overview
+
+The Campaign Assessment System provides users with encouraging, actionable feedback about their campaign development progress. Instead of showing intimidating numerical scores, it uses descriptive RPG-themed states that celebrate growth and provide clear next steps.
+
```
<3

aniham commented on docs/ASSESSMENT_SYSTEM.md:
```diff
@@ -0,0 +1,296 @@
+
+### Core Components
+
+1. **Assessment Service** (`src/services/assessment-service.ts`)
+   - Analyzes user and campaign state
+   - Calculates readiness scores
+   - Generates recommendations
+
+2. **Campaign State Utilities** (`src/lib/campaign-state-utils.ts`)
+   - Converts scores to descriptive states
+   - Generates encouraging summaries
+   - Provides next milestone guidance
+
+3. **Assessment Types** (`src/types/assessment.ts`)
+   - Centralized type definitions
+   - Eliminates duplication across codebase
+
+4. **Assessment Tools** (`src/tools/campaign-context/assessment-tools.ts`)
+   - Tool wrappers for agent access
+   - Campaign-specific assessments
```
did you want this here or did cursor put it in? i hate this stuff because it gets out of sync with file names almost immediately. cursor is not good at updating the shit it started. and then it confuses itself

ofisk replied:
I asked it to update docs. interested in the point though, if you can elaborate more. are you saying any docs that name specific files in general are at risk here?

aniham replied:
yes :)

aniham commented on docs/ASSESSMENT_SYSTEM.md:
```diff
@@ -0,0 +1,296 @@
+
+Scores are calculated based on three dimensions:
+
+1. **Campaign Context** (0-50 points)
+   - 0 items: 10 points + priority area
+   - 1-2 items: 30 points + priority area
+   - 3+ items: 50 points
+
+2. **Character Development** (0-50 points)
+   - 0 characters: 10 points + priority area
+   - 1-2 characters: 30 points + priority area
+   - 3+ characters: 50 points
+
+3. **Resources** (0-40 points)
+   - 0 resources: 10 points + priority area
+   - 1-4 resources: 30 points + priority area
+   - 5+ resources: 40 points
+
```
this is cool. we can add weights too i think, later on!!

ofisk replied:
I thought so too. This is something on the mind in general for me. I'm not sure what exactly it should mean and/or how to calculate it. For example, it seems unhelpful to like count the number of vectors or something, but I assume there's some correlation there for readiness of a given "shard". Outside of that, what's smart to do here? We can home brew some algorithm like this but it's really subjective and a "secret sauce" area that I'm sure we could tweak forever and ever and never really get fully right. Will be fun to play with nonetheless.

aniham replied:
> I'm not sure what exactly it should mean and/or how to calculate it

same :)

aniham replied:
> never really get fully right

dk about that. i think we'll get better and better to where "fully" is irrelevant

aniham replied:
> count the number of vectors or something

i think it's actually going to evolve more to interactions between the data and how well it's connected. not "did you update a file" but "what part of your file is covered in your campaign plan?" or "how many characters from your character list have a story line inside the campaign?"

aniham commented on docs/ASSESSMENT_SYSTEM.md:
```diff
@@ -0,0 +1,296 @@
+
+### Progressive Milestones
+
+Each state has tailored guidance:
+
+- **Fresh Start → Newly Forged**: Focus on creating first elements (character, resource, location)
+- **Newly Forged → Taking Root**: Add diversity (2-3 of each type)
+- **Taking Root → Taking Shape**: Develop narrative connections and detail
+- **Taking Shape → Growing Strong**: Deepen relationships and world complexity
+- **Growing Strong → Flourishing**: Add variety and atmospheric richness
+- **Flourishing → Well-Traveled**: Create complex systems and lore
+- **Well-Traveled → Epic-Ready**: Polish with interconnected content
+- **Epic-Ready → Legendary**: Add advanced custom content
+- **Legendary+**: Maintain excellence and focus on session planning
+
```
like. idk. if you want this here, we'll keep it. i feel like this should be documented in the prompt guidance itself. you know?

ofisk replied:
Should be there as well. Specifically mean the actionable suggestions from your first comment. I'm ok with removing it from docs. I guess the only advantage is in having a one stop shop to read up on how the app is doing assessment.

aniham commented on docs/ASSESSMENT_SYSTEM.md:
````diff
@@ -0,0 +1,296 @@
+
+```typescript
+export function getCampaignState(score: number): string {
+  if (score >= 90) return "Legendary";
````
had a comment in the other PR too - i see this switch statement (that's not a switch) a lot :)

aniham commented on docs/ASSESSMENT_SYSTEM.md:
````diff
@@ -0,0 +1,296 @@
+**generateReadinessSummary()**
+
+```typescript
+export function generateReadinessSummary(
````
should just be in the code

aniham commented on src/lib/campaign-state-utils.ts:
```diff
@@ -0,0 +1,208 @@
+ * Get next milestone for campaign progression
+ * Provides specific, actionable steps to reach the next state
+ */
+export function getNextMilestone(score: number): {
```
this method looks good. i assume we'll replace the more deterministic text with variants that the agent comes up with later on?

ofisk replied:
would love to do that, especially given how many different ways you could "advance" the context of a campaign.

aniham reviewed (approved):
this looks great actually. i had some of the same comments about disagreeing with cursor on what's good documentation and what's good visuals. but i already noted those in the other pr. this is great and good to go

some questions (just one ig): how do you see the scoring and the knowledge going forward? what do we grow? for example, we have tyings we score and focus and and recommend? how do we expand those from leaning more deterministic to smarter and smarter?

ofisk committed:
Docs: Add assessment system documentation and improve inline comments

- Document scoring algorithm in getCampaignReadiness method
- Clarify tool purposes with enhanced descriptions
- Update assessment tool comments for better UX clarity
- Fix linter warnings in tests and utilities
